src/middlewared/middlewared/plugins/cache.py

```diff
@@ -26,18 +27,36 @@ def get(self, key):
         Raises:
             KeyError: not found in the cache
         """
+        if isinstance(self.__cache[key][-1], datetime):
+            self.get_timeout(key)
+            return self.__cache[key][0]
+
         return self.__cache[key]
 
-    @accepts(Str('key'), Any('value'))
-    def put(self, key, value):
+    @accepts(Str('key'), Any('value'), Datetime('timeout', required=False))
+    def put(self, key, value, timeout):
         """
         Put `key` of `value` in the cache.
         """
-        self.__cache[key] = value
+        if timeout is not None:
```

Comment:
It's very inconvenient (and will become much worse). You can create a namedtuple `CacheElement = namedtuple("CacheElement", ["value", "timeout"])` and store it in `self.__cache` instead